src/Reopt/TypeInference/ConstraintGen.hs

```diff
@@ -246,6 +215,33 @@ warn s = CGenM $ warnings <>= [Warning s]
 freshTyVar :: CGenM ctx arch TyVar
 freshTyVar = CGenM $ TyVar <$> (nextFreeTyVar <<+= 1)
 
+-- | Returns a fresh row var. N.B. IMPORTANT: this should be used for any record types
+-- in constraints to represent possible additional fields. And... remove `_` prefix
+-- when doing so =)
+_freshRowVar :: CGenM ctx arch RowVar
+_freshRowVar = CGenM $ RowVar <$> (nextFreeRowVar <<+= 1)
+
+-- | Converts the limited `Constraint` grammar to the more general
+-- `TyConstraint` type used for constraint solving.
+tyConstraint :: Constraint -> CGenM ctx arch TyConstraint
+tyConstraint = \case
+  CEq t1 t2 -> pure $ eqTC t1 t2
+  CAddrWidthAdd ret lhs rhs -> maybePtrBinop ret lhs rhs
+  CAddrWidthSub ret lhs rhs -> maybePtrBinop ret lhs rhs
+  CBVNotPtr t -> pure $ eqTC t NumTy
+  CIsPtr t -> do
+    x <- UnknownTy <$> freshTyVar
+    pure $ eqTC t (PtrTy x)
+  where
+    maybePtrBinop ret lhs rhs = do
+      x1 <- UnknownTy <$> freshTyVar
+      x2 <- UnknownTy <$> freshTyVar
+      pure $ orTC
+        [ andTC [eqTC ret NumTy,      eqTC lhs NumTy,      eqTC rhs NumTy]
+        , andTC [eqTC ret (PtrTy x2), eqTC lhs (PtrTy x1), eqTC rhs NumTy]
+        , andTC [eqTC ret (PtrTy x2), eqTC lhs NumTy,      eqTC rhs (PtrTy x1)]
+        ]
```

Comment:
@pnwamk 
Is there a reason to create two type variables here? When would they not equate each other?

Reply:
It seemed safer to assume pointer arithmetic doesn't _always_ result in a pointer to a value of the same type. Is it sound to assume otherwise?

Reply:
Hmm maybe I'm thinking too much in C terms rather than wacky assembly wild west terms! :-D

Reply:
They will only be the same if you are doing array ops, for field indexing they will be differnt